File: scripts/split_lectio.py
#!/usr/bin/env python3
"""Split long PL units into lectio-sized chunks.

latin-units.json is left alone. Chunks are derived after English is attached,
and written onto de-gradibus.json as a `chunks` array.

IDs for a unit broken into three parts look like praefatio-s1, praefatio-s2,
praefatio-s3.
"""
from __future__ import annotations

import json
import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
OUT = ROOT / "src/content/gradibus/de-gradibus.json"

TARGET_WORDS = 60
NEVER_SPLIT = {"title", "chapter-title"}

# Period after these tokens is not a sentence end (citations, n. 1, S. Bernardus).
ABBREV = {
    "al", "apoc", "bar", "cant", "cantic", "cap", "cf", "col", "cor", "dan",
    "deut", "dr", "eccl", "eccli", "eph", "etc", "exod", "ezech", "ezek",
    "gal", "gen", "hab", "heb", "hebr", "id", "ibid", "ioan", "isa", "jac",
    "jer", "jn", "job", "joel", "john", "jos", "jud", "lev", "lk", "luc",
    "mal", "mar", "marc", "matt", "matth", "mic", "mk", "mr", "mt", "n",
    "nah", "no", "num", "os", "paral", "pet", "petr", "phil", "philipp",
    "prov", "ps", "psal", "reg", "rev", "rom", "s", "sap", "scil", "seq",
    "sir", "song", "sq", "ss", "st", "thess", "tim", "tit", "tob", "tom",
    "v", "viz", "vs", "vv", "wis", "zach",
}
MIN_SENT_WORDS = 12
MIN_CHUNK_WORDS = 20


def _is_abbrev(token: str) -> bool:
    stripped = token.strip("«»“”\"'").rstrip(".")
    if len(stripped) == 1 and stripped.isalpha():
        return True
    return stripped.casefold() in ABBREV


def split_sentences(text: str, *, allow_lowercase: bool = False) -> list[str]:
    """Split on .?! skipping citation abbreviations. Parentheses are not a fence:
    PL asides are often unbalanced and would swallow the rest of a unit.

    O'Donnell-style all-lowercase Latin does not capitalize after a period, so
    pass allow_lowercase=True for those texts. Capitalized Latin (Bernard) keeps
    the default, which still requires an uppercase letter after .?!
    """
    text = text.strip()
    if not text:
        return []

    sentences: list[str] = []
    buf: list[str] = []
    i = 0
    n = len(text)

    while i < n:
        ch = text[i]
        if ch in ".?!":
            buf.append(ch)
            j = i + 1
            while j < n and text[j] in "\"'»”’)":
                buf.append(text[j])
                j += 1
            while j < n and text[j].isspace():
                j += 1
            at_end = j >= n
            nxt = text[j] if j < n else ""
            core = re.sub(r'[.?!]+["\'»”’)]*$', "", "".join(buf))
            last_m = re.search(r"([A-Za-z]+)$", core)
            last = last_m.group(1) if last_m else ""
            skip = _is_abbrev(last) or nxt.isdigit()
            letter_start = nxt.isalpha() if allow_lowercase else nxt.isupper()
            starts = at_end or (nxt and (letter_start or nxt in "«“\"'"))
            if starts and not skip:
                sent = "".join(buf).strip()
                if sent:
                    sentences.append(sent)
                buf = []
                i = j
                continue
            i += 1
            continue
        buf.append(ch)
        i += 1

    tail = "".join(buf).strip()
    if tail:
        sentences.append(tail)
    return sentences or [text]


def word_count(text: str) -> int:
    return len(text.split()) or 1


def coalesce(sents: list[str], min_words: int = MIN_SENT_WORDS) -> list[str]:
    """Attach very short sentences to a neighbor so lectio pages are not one word."""
    if not sents:
        return sents
    out = [sents[0]]
    for sent in sents[1:]:
        if word_count(sent) < min_words:
            out[-1] = join_sents([out[-1], sent])
        else:
            out.append(sent)
    if len(out) >= 2 and word_count(out[0]) < min_words:
        out[0:2] = [join_sents(out[0:2])]
    return out


def desired_parts(n_sents: int, words: int, kind: str) -> int:
    if kind in NEVER_SPLIT:
        return 1
    if words <= int(TARGET_WORDS * 1.25) or n_sents <= 1:
        return 1
    return max(1, min(n_sents, round(words / TARGET_WORDS)))


def partition(items: list[str], n_parts: int) -> list[list[str]]:
    """Split `items` into n_parts runs, keeping word counts as even as possible."""
    n_parts = max(1, min(n_parts, len(items)))
    if n_parts == 1:
        return [items]

    weights = [word_count(item) for item in items]
    m = len(items)
    inf = 10**9
    prefix = [0]
    for w in weights:
        prefix.append(prefix[-1] + w)

    dp = [[inf] * (n_parts + 1) for _ in range(m + 1)]
    back = [[0] * (n_parts + 1) for _ in range(m + 1)]
    dp[0][0] = 0
    for i in range(1, m + 1):
        for k in range(1, min(n_parts, i) + 1):
            for j in range(k - 1, i):
                cost = max(dp[j][k - 1], prefix[i] - prefix[j])
                if cost < dp[i][k]:
                    dp[i][k] = cost
                    back[i][k] = j

    groups: list[list[str]] = []
    i, k = m, n_parts
    while k:
        j = back[i][k]
        groups.append(items[j:i])
        i, k = j, k - 1
    groups.reverse()
    return groups


def merge_tiny(groups: list[list[str]], min_words: int = MIN_CHUNK_WORDS) -> list[list[str]]:
    """Fold leftover scraps into the previous lectio part."""
    if not groups:
        return groups
    out = [groups[0]]
    for group in groups[1:]:
        if word_count(join_sents(group)) < min_words:
            out[-1] = out[-1] + group
        else:
            out.append(group)
    if len(out) >= 2 and word_count(join_sents(out[0])) < min_words:
        out[0:2] = [out[0] + out[1]]
    return out


def join_sents(sents: list[str]) -> str:
    return " ".join(sents).strip()


def chunk_unit(unit: dict, *, allow_lowercase: bool = False) -> list[dict]:
    latin = unit["latin"]
    english = unit.get("english") or ""
    # Latin may be all-lowercase (O'Donnell); English renderings stay capitalized.
    la = coalesce(split_sentences(latin, allow_lowercase=allow_lowercase))
    en = coalesce(split_sentences(english)) if english else [""]
    n = desired_parts(len(la), word_count(latin), unit.get("kind", ""))
    n = max(1, min(n, len(la), max(1, len(en))))

    la_groups = merge_tiny(partition(la, n))
    en_groups = partition(en, len(la_groups))
    while len(en_groups) < len(la_groups) and len(la_groups) > 1:
        la_groups[-2] = la_groups[-2] + la_groups[-1]
        la_groups.pop()
    parts = len(la_groups)
    chunks = []
    for i, (la_g, en_g) in enumerate(zip(la_groups, en_groups), start=1):
        chunk_id = unit["id"] if parts == 1 else f"{unit['id']}-s{i}"
        chunks.append({
            "id": chunk_id,
            "part": i,
            "parts": parts,
            "latin": join_sents(la_g),
            "english": join_sents(en_g),
        })
    return chunks


def add_chunks(units: list[dict]) -> list[dict]:
    for unit in units:
        unit["chunks"] = chunk_unit(unit)
    return units


def report(units: list[dict]) -> None:
    total = 0
    for unit in units:
        chunks = unit["chunks"]
        total += len(chunks)
        if len(chunks) == 1:
            continue
        words = " + ".join(str(word_count(c["latin"])) for c in chunks)
        print(f"{unit['id']}: {len(chunks)} parts ({words} words)")
    print(f"\n{len(units)} source units -> {total} lectio units")


def main() -> None:
    path = Path(sys.argv[1]) if len(sys.argv) > 1 else OUT
    units = json.loads(path.read_text(encoding="utf-8"))
    add_chunks(units)
    path.write_text(json.dumps(units, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    report(units)
    print(f"wrote chunks -> {path}")


if __name__ == "__main__":
    main()

File: src/content/splitLectio.ts
import type { LectioChunk } from "../types";

/** Same target as scripts/split_lectio.py (Gradibus lectio pages). */
const TARGET_WORDS = 60;
const NEVER_SPLIT = new Set(["title", "chapter-title"]);
const MIN_SENT_WORDS = 12;
const MIN_CHUNK_WORDS = 20;
const ABBREV = new Set([
  "al",
  "apoc",
  "act",
  "bar",
  "baruch",
  "cant",
  "cantic",
  "cap",
  "cf",
  "col",
  "coloss",
  "colossens",
  "cor",
  "dan",
  "deut",
  "dr",
  "eccl",
  "eccle",
  "eccli",
  "eph",
  "ephes",
  "esdr",
  "etc",
  "exod",
  "ezech",
  "ezek",
  "gal",
  "galat",
  "gen",
  "hab",
  "habac",
  "heb",
  "hebr",
  "iob",
  "id",
  "ibid",
  "ioan",
  "ioel",
  "isa",
  "isai",
  "iudic",
  "ierem",
  "iac",
  "jac",
  "jer",
  "jn",
  "job",
  "joel",
  "john",
  "jos",
  "jud",
  "lev",
  "levit",
  "lk",
  "luc",
  "mal",
  "malach",
  "mar",
  "marc",
  "math",
  "matt",
  "matth",
  "mach",
  "machab",
  "mic",
  "mich",
  "mk",
  "mr",
  "mt",
  "n",
  "nah",
  "no",
  "num",
  "os",
  "ose",
  "ovid",
  "paral",
  "pet",
  "petr",
  "phil",
  "philipp",
  "philip",
  "philem",
  "prov",
  "ps",
  "psal",
  "reg",
  "rev",
  "rom",
  "ruth",
  "s",
  "sap",
  "scil",
  "seq",
  "serm",
  "sir",
  "song",
  "soph",
  "sophon",
  "sq",
  "ss",
  "st",
  "thess",
  "thren",
  "tim",
  "tit",
  "tob",
  "tom",
  "v",
  "viz",
  "vs",
  "vv",
  "wis",
  "zach",
]);
const CLOSING = new Set(["\"", "'", "»", "”", "’", ")"]);
const OPENING = new Set(["«", "“", "\"", "'"]);

export type Chunkable = {
  id: string;
  latin: string;
  english: string;
  kind?: string;
};

function isAbbrev(token: string): boolean {
  const stripped = token.replace(/[«»“”"'']/g, "").replace(/\.+$/, "");
  if (stripped.length === 1 && /[A-Za-z]/.test(stripped)) return true;
  return ABBREV.has(stripped.toLowerCase());
}

function splitSentences(text: string, allowLowercase: boolean): string[] {
  const source = text.trim();
  if (!source) return [];

  const sentences: string[] = [];
  let buf = "";
  let i = 0;
  const n = source.length;

  while (i < n) {
    const ch = source[i];
    if (ch === "." || ch === "?" || ch === "!") {
      buf += ch;
      let j = i + 1;
      while (j < n && CLOSING.has(source[j])) {
        buf += source[j];
        j += 1;
      }
      while (j < n && /\s/.test(source[j])) j += 1;
      const atEnd = j >= n;
      const nxt = atEnd ? "" : source[j];
      const core = buf.replace(/[.?!]+["'»”’)]*$/, "");
      const lastM = core.match(/([A-Za-z]+)$/);
      const last = lastM?.[1] ?? "";
      const skip = isAbbrev(last) || /\d/.test(nxt);
      const letterStart = nxt
        ? allowLowercase
          ? /[A-Za-z]/.test(nxt)
          : nxt === nxt.toUpperCase() && /[A-Za-z]/.test(nxt)
        : false;
      const starts = atEnd || (nxt && (letterStart || OPENING.has(nxt)));
      if (starts && !skip) {
        const sent = buf.trim();
        if (sent) sentences.push(sent);
        buf = "";
        i = j;
        continue;
      }
      i += 1;
      continue;
    }
    buf += ch;
    i += 1;
  }

  const tail = buf.trim();
  if (tail) sentences.push(tail);
  return sentences.length > 0 ? sentences : [source];
}

function wordCount(text: string): number {
  const n = text.trim().split(/\s+/).filter(Boolean).length;
  return n || 1;
}

function joinSents(sents: string[]): string {
  return sents.join(" ").trim();
}

function coalesce(sents: string[], minWords = MIN_SENT_WORDS): string[] {
  if (sents.length === 0) return sents;
  const out = [sents[0]];
  for (const sent of sents.slice(1)) {
    if (wordCount(sent) < minWords) {
      out[out.length - 1] = joinSents([out[out.length - 1], sent]);
    } else {
      out.push(sent);
    }
  }
  if (out.length >= 2 && wordCount(out[0]) < minWords) {
    out.splice(0, 2, joinSents(out.slice(0, 2)));
  }
  return out;
}

function desiredParts(nSents: number, words: number, kind: string): number {
  if (NEVER_SPLIT.has(kind)) return 1;
  if (words <= Math.floor(TARGET_WORDS * 1.25) || nSents <= 1) return 1;
  return Math.max(1, Math.min(nSents, Math.round(words / TARGET_WORDS)));
}

function partition(items: string[], nParts: number): string[][] {
  nParts = Math.max(1, Math.min(nParts, items.length));
  if (nParts === 1) return [items];

  const weights = items.map(wordCount);
  const m = items.length;
  const inf = 1e9;
  const prefix = [0];
  for (const w of weights) prefix.push(prefix[prefix.length - 1] + w);

  const dp = Array.from({ length: m + 1 }, () => Array(nParts + 1).fill(inf));
  const back = Array.from({ length: m + 1 }, () => Array(nParts + 1).fill(0));
  dp[0][0] = 0;
  for (let i = 1; i <= m; i++) {
    for (let k = 1; k <= Math.min(nParts, i); k++) {
      for (let j = k - 1; j < i; j++) {
        const cost = Math.max(dp[j][k - 1], prefix[i] - prefix[j]);
        if (cost < dp[i][k]) {
          dp[i][k] = cost;
          back[i][k] = j;
        }
      }
    }
  }

  const groups: string[][] = [];
  let i = m;
  let k = nParts;
  while (k) {
    const j = back[i][k];
    groups.push(items.slice(j, i));
    i = j;
    k -= 1;
  }
  groups.reverse();
  return groups;
}

function mergeTiny(groups: string[][], minWords = MIN_CHUNK_WORDS): string[][] {
  if (groups.length === 0) return groups;
  const out = [groups[0]];
  for (const group of groups.slice(1)) {
    if (wordCount(joinSents(group)) < minWords) {
      out[out.length - 1] = out[out.length - 1].concat(group);
    } else {
      out.push(group);
    }
  }
  if (out.length >= 2 && wordCount(joinSents(out[0])) < minWords) {
    out.splice(0, 2, out[0].concat(out[1]));
  }
  return out;
}

/**
 * Break a Latin+English unit into Gradibus-style lectio pages (~60 words).
 * O'Donnell-style all-lowercase Latin needs `{ allowLowercase: true }`.
 * English is always split with the capitalized-sentence rule (Pusey, etc.).
 *
 * `sentenceAligned` (used by cantica) assumes the English was written with
 * one sentence per Latin sentence, in order. Pages are then cut on the Latin
 * sentence boundaries and the English is sliced with the exact same
 * sentence-index boundaries, so every page's Latin and English are the same
 * material regardless of how word lengths fall. All other callers keep the
 * legacy independent word-balance grouping below.
 */
export function chunkLectio(
  unit: Chunkable,
  opts: { allowLowercase?: boolean; sentenceAligned?: boolean } = {},
): LectioChunk[] {
  const allowLowercase = opts.allowLowercase ?? false;
  const english = unit.english || "";

  const laRaw = splitSentences(unit.latin, allowLowercase);
  const enRaw = english ? splitSentences(english, false) : [];

  if (
    opts.sentenceAligned === true &&
    laRaw.length > 0 &&
    enRaw.length === laRaw.length
  ) {
    let n = desiredParts(laRaw.length, wordCount(unit.latin), unit.kind ?? "");
    n = Math.max(1, Math.min(n, laRaw.length, Math.max(1, enRaw.length)));
    const laGroups = mergeTiny(partition(laRaw, n));
    const sizes = laGroups.map((g) => g.length);
    const enGroups: string[][] = [];
    let off = 0;
    for (const s of sizes) {
      enGroups.push(enRaw.slice(off, off + s));
      off += s;
    }
    const parts = laGroups.length;
    return laGroups.map((laG, index) => {
      const part = index + 1;
      return {
        id: parts === 1 ? unit.id : `${unit.id}-s${part}`,
        part,
        parts,
        latin: joinSents(laG),
        english: joinSents(enGroups[index] ?? [""]),
      };
    });
  }

  const la = coalesce(laRaw);
  const en = english ? coalesce(enRaw) : [""];
  let n = desiredParts(la.length, wordCount(unit.latin), unit.kind ?? "");
  n = Math.max(1, Math.min(n, la.length, Math.max(1, en.length)));

  let laGroups = mergeTiny(partition(la, n));
  const enGroups = partition(en, laGroups.length);
  while (enGroups.length < laGroups.length && laGroups.length > 1) {
    laGroups[laGroups.length - 2] = laGroups[laGroups.length - 2].concat(
      laGroups[laGroups.length - 1],
    );
    laGroups.pop();
  }
  const parts = laGroups.length;
  return laGroups.map((laG, index) => {
    const part = index + 1;
    return {
      id: parts === 1 ? unit.id : `${unit.id}-s${part}`,
      part,
      parts,
      latin: joinSents(laG),
      english: joinSents(enGroups[index] ?? [""]),
    };
  });
}
